design/one-pager-class-selection.md

```diff
@@ -0,0 +1,371 @@
+  and used by applications running in a distinct _application namespace_ such as
+  `widget-service` or `acme-team`.
+
+Note that two forms (classes?) of resource class currently exist; see [portable
+resource classes] for details.
+
+## Goals
+
+The goal of this proposal is to __make it easier for potential Crossplane users
+to learn and take advantage of the portability and separation of concerns__ that
+are enabled by dynamic provisioning. A _secondary_ goal is to maintain the
+flexibility and power that Crossplane offers complex organisations with many
+application and infrastructure namespaces.
+
+## Proposal
+
+This document proposes Crossplane retire the concept of portable resource
+classes. Resource claims would instead be matched directly to non-portable
+resource classes (referred to henceforth as simply 'resource classes') using
+[label selectors].
+
+```yaml
+apiVersion: database.crossplane.io/v1alpha1
+kind: PostgreSQLInstance
+metadata:
+  namespace: acme-team
+  name: gitlab-database
+spec:
+  classSelector:
+    matchLabels:
+      stack: gitlab
+      grade: experimental
+      region: us-east-1
+```
+
+Consider the hypothetical resource claim above, created by an equally
+hypothetical GitLab [stack]. It declares a need for an experimental grade
+PostgreSQL instance in region us-east-1 that is suitable for use with the GitLab
+stack. Any compatible resource class (whether it be an `RDSInstanceClass` or a
+`CloudSQLInstanceClass`) matching the resource claim's labels would be eligible
+to satisfy it. If more than one such class existed one would be selected at
+random. A resource claim that omitted `.spec.classSelector`, `.spec.classRef`,
+and `.spec.resourceRef` would be deemed to be completely unopinionated and would
+match a compatible resource class at random.
+
+This pattern is likely familiar to Kubernetes and to Crossplane users - it's
+used to select which nodes a pod may run on, which pods are endpoints of a
+service, and which Kubernetes cluster a Crossplane workload may run on.
+
+### Matching Scope
+
+Resource classes would be selected only from within the same namespace as the
+resource claim by default; this is the simplest, safest, and least surprising
+default for small scale Crossplane deployments. The infrastructure namespace
+pattern would be enabled by creating an association between application and
+infrastructure namespaces using namespace level annotations. For example:
+
+```yaml
+apiVersion: v1
+kind: Namespace
+metadata:
+  name: acme-team
+  annotations:
+    resource-class.crossplane.io/select-namespaces: "acme-team, experimental-infra"
+```
+
+The above `acme-team` application namespace declares that any resource claims
+created within it should be matched against resource classes from namespaces
+`acme-team` or `experimental-infra`. Note that when the `select-namespaces`
+annotation exists it overrides the default behaviour of selecting only within
```

Comment:
After finishing my read the doc, it appears that without some sort of weighting implemented, then there is really no such thing as a traditional `default` anymore. This is in-line with this comment https://github.com/crossplaneio/crossplane/issues/870#issuecomment-537396943

> Strongly suggesting that by convention resource claims with no opinion use a label selector like `{"default": "true"}`

So I guess my question above boils down to how do we handle the scenario where there are multiple classes of the same `kind` that match labels but are in different namespaces that have been specified in the `select-namespaces` annotation. The `Technical Implementation` section below covers the case that two separate controllers both find a satisfactory class (race condition), but not where a single controller finds multiple satisfactory classes in different namespaces.

Reply:
> How do we handle the scenario where there are multiple classes of the same kind that match labels but are in different namespaces that have been specified in the select-namespaces annotation.

We pick one at random. The selection process within the global set of matching resource classes is effectively:

1. One of each kind of eligible resource class is picked at random.
1. One of the kinds randomly selected in class one is picked at random.

I'd argue that this means that it's still _possible_ to support the "defaulting" pattern that we do today - but doing so requires a convention to be established when writing resource classes and claims. Resource claim authors that wanted the default would have to select on some label (i.e. `default: "true"), and resource class authors would have to ensure only one class was labelled the default within each matching scope. The latter constraint exists today - you may only label one portable resource class as the default. The issue is what happens when multiple classes are labelled as the default - under this proposal you get a randomly selected default whereas today you get an error. I do see that this could be a [footgun](https://en.wiktionary.org/wiki/footgun), but I don't think it's any more a footgun than existing instances of label selection and scheduling within Kubernetes.

I do wonder how many folks will still choose to use simple defaulting when they have a slightly more expressive language to describe things like "I want the default for gitlab", or similar.